src/algebra/group/unique_prods.lean

```diff
@@ -0,0 +1,213 @@
+/-
+Copyright (c) 2022 Damiano Testa. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Damiano Testa
+-/
+import data.finset.preimage
+
+/-!
+#  Unique products and related notions
+
+A group `G` has *unique products* if for any two non-empty finite subsets `A, B ⊂ G`, there is an
+element `g ∈ A * B` that can be written uniquely as a product of an element of `A` and an element
+of `B`.  We call the formalization this property `unique_prods`.  Since the condition requires no
+property of the group operation, we define it for a Type simply satisfying `has_mul`.  We also
+introduce the analogous "additive" companion, `unique_sums` and link the two so that `to_additive`
+converts `unique_prods` into `unique_sums`.
+
+Here you can see several examples of Types that have `unique_sums/prods`
+(`apply_instance` uses `covariants.to_unique_prods` and `covariants.to_unique_sums`).
+```lean
+import data.real.basic
+
+example : unique_sums ℕ   := by apply_instance
+example : unique_sums ℕ+  := by apply_instance
+example : unique_sums ℤ   := by apply_instance
+example : unique_sums ℚ   := by apply_instance
+example : unique_sums ℝ   := by apply_instance
+example : unique_prods ℕ+ := by apply_instance
+```
+
+A Type that does not have `unique_prods`.
+```lean
+example : ¬ unique_prods ℕ :=
```

Comment:
This should go to `counterexamples/`.

Reply:
I added a TODO to move the counterexamples: #16236 already creates a counterexample files where these two would fit well and I prefer to avoid creating dependencies among PRs that are really independent.  Once #16236 gets merged, I will move these examples to their natural place.

Does this seem like a good compromise?